File: 2023/day1/Day1.kt
import java.nio.file.Files
import java.nio.file.Path

fun main() {
    val lines = Files.readAllLines(Path.of("src/main/resources/in.txt"))

    var sumP1 = 0
    var sumP2 = 0

    val pattern = Regex("(?=(\\d|one|two|three|four|five|six|seven|eight|nine))")

    for (line in lines) {
        val matches = pattern.findAll(line).map { it.groupValues[1] }.toList()

        val matchesP1 = matches.filter { it.toIntOrNull() != null }
        sumP1 += (matchesP1.first() + matchesP1.last()).toInt()

        val matchesP2 = matches.map {
            it.replace("one", "1")
                .replace("two", "2")
                .replace("three", "3")
                .replace("four", "4")
                .replace("five", "5")
                .replace("six", "6")
                .replace("seven", "7")
                .replace("eight", "8")
                .replace("nine", "9")
        }
        sumP2 += (matchesP2.first() + matchesP2.last()).toInt()
    }

    println("Part 1: $sumP1")
    println("Part 2: $sumP2")
}

File: 2023/day1/day1.py
import regex as re

if __name__ == '__main__':
    with open('in.txt') as file:
        lines = file.read().splitlines()

    sum_p1 = 0
    sum_p2 = 0

    for line in lines:
        matches = re.findall(r'\d|one|two|three|four|five|six|seven|eight|nine', line, overlapped=True)

        matches_p1 = list(filter(lambda d: d.isdigit(), matches))
        sum_p1 += int(matches_p1[0] + matches_p1[-1])

        matches_p2 = list(map(lambda m: m.replace('one', '1')
                              .replace('two', '2')
                              .replace('three', '3')
                              .replace('four', '4')
                              .replace('five', '5')
                              .replace('six', '6')
                              .replace('seven', '7')
                              .replace('eight', '8')
                              .replace('nine', '9'), matches))
        sum_p2 += int(matches_p2[0] + matches_p2[-1])

    print('Part 1:', sum_p1)
    print('Part 2:', sum_p2)
